common/djangoapps/util/keyword_substitution.py

```diff
@@ -53,30 +46,36 @@ def substitute_keywords(string, user=None, course=None):
     the loop below, and reduces the need for piling up if elif else statements
     when the keyword pool grows.
     """
-    if user is None or course is None:
-        # Cannot proceed without course and user information
-        return string
 
-    for key, func in KEYWORD_FUNCTION_MAP.iteritems():
+    # do this lazily to avoid unneeded database hits
+    KEYWORD_FUNCTION_MAP = {
+        '%%USER_ID%%': lambda: anonymous_id_from_user_id(user_id),
+        '%%USER_FULLNAME%%': lambda: context['name'],
+        '%%COURSE_DISPLAY_NAME%%': lambda: context['course_title'],
+        '%%COURSE_END_DATE%%': lambda: context.get('course_end_date'),
```

Comment:
I did this because before we even call this function, we're testing to see if `"name"` and `"course_title"` are in the email context. But for aesthetic reasons, I could change it to `.get`